% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/MainFunRun.R
\name{OmicsPLMM}
\alias{OmicsPLMM}
\title{Multi-kernel linear mixed model with adaptive lasso for prediction analysis on high-dimensional multi-omics data}
\usage{
OmicsPLMM(InputData = list(), OmicsData = list(),
  OmicsDataMap = list(), trainID = NULL, OmicsKernelMatrix = list(),
  annotation = NULL, Y = NULL, X = NULL, kernelsOmics = NA,
  AllRegions = 0, stdData = TRUE, predict = FALSE,
  weight.fixed = NULL, weight.random = NULL, maxiter = 1000,
  outputall = 0)
}
\arguments{
\item{InputData}{A \code{OmicPredData} class object that is generated by \code{ReadOmicPLMM}. If provided, will ignore \code{OmicsData},\code{OmicsDataMap},\code{trainID},\code{OmicsKernelMatrix},\code{annotation},\code{X},\code{Y},\code{kernelsOmics},\code{AllRegion} \code{stdData}.}

\item{OmicsData}{A list with each element representing data from one omic. Each row is an observation, and each column represents a predictor. NA represents missing. For the subject with missing omics,the subject should be in the corresponding omic, but with missing entry.}

\item{OmicsDataMap}{A list with each element representing annotation for one omic. Each row is a predictor. Each row should have at least 2 columns ("chromosome","position"). The number of rows should be the same as the number of columns in the corresponding omic data.}

\item{trainID}{Subject ID for the training individuals.}

\item{OmicsKernelMatrix}{A list with each element represents similarity matrices for each omic data. If provided, will ignore the inputs from \code{OmicsData} and \code{OmicsDataMap}.}

\item{annotation}{A data frame providing information to cut the genome. It has at least four columns: gene, chr, start, end. The gene column must be unique. The other three columns contain information about chromosome, start and end of each regomic region. #####If Kinship is not provided, then annotation must be provided.}

\item{Y}{A vector of phenotypes with each name being subject ID. If provided, will ignore \code{phenofile}.}

\item{X}{A matrix of demographic variables, should be of the same order as Y (i.e. \code{rownames(X)=names(Y)}). The intercept column is not needed.}

\item{kernelsOmics}{kernels used for other omic data. Currently, it takes linear kernel for non-genomic data. For genomic data, it can take linear and IBS kernels.  Its length should be the same to the number of omics data if provided.}

\item{stdData}{A bool to indicate whether data should be centered and scaled with sd=1. Default=TRUE.}

\item{predict}{A bool indicating if prediction is needed. Default is FALSE.}

\item{weight.fixed}{Weight that one wants to assign to covariate. Default is NULL.}

\item{weight.random}{Weight that one wants to assign to random effect. This should be as the same number of random effects included in the model. Suggest to give NULL, and default is NULL.}

\item{maxiter}{Maximum number of iterations. Default is 1000.}

\item{outputall}{Controls the details of output. \code{outputall=0}, only the prediction and original outputs are saved.  \code{outputall=1}, in addition to the above, the model fit is also saved. \code{outputall=2}, in addition to the above, the similarity matrix is also save (can be large, not suggested).}

\item{AllRegion}{A bool to indicate whether GSM should be calculated from the entire genome. Default=0.}
}
\value{
A list that contains original output and the predicted value (if any). The other output is controlled by \code{outputall}.
}
\description{
\code{OmicsPLMM} prepares the data for analyses
}
\details{
This function is used to pre-process the data used by the Multi-Kernel Linear Mixed Model with Adaptive Lasso method.
}
\examples{
## Using this function to prepare teh data for the analysis ##
OmicsData=list();  OmicsDataMap=list();OmicsKernelMatrix=list();
OmicsData$Genomic=GenoEg1;
OmicsDataMap$Genomic=GenoMapEg1;
Y=YEg1;
XCov=ExprEg1;
trainID=trainIDEg1;
predict=TRUE;
weight.fixed=weight.random=NULL;
maxiter=1000;
outputall=0;
ReadData<-ReadOmicPLMM(OmicsData=OmicsData,  OmicsDataMap=OmicsDataMap, trainID=trainID,OmicsKernelMatrix=list(), annotation=annotation1, Y=Y, X=Expr, kernelsOmics =NA, AllRegions=0)

## Using prepared Data ##
ResultTest1=OmicsPLMM(InputData=ReadData,
predict = predict, weight.fixed = weight.fixed, weight.random = weight.random,
maxiter = maxiter, outputall = outputall)
cor(ResultTest1$Ypred,ResultTest1$outcome[!names(ResultTest1$outcome) \%in\% trainID])

ResultTest=OmicsPLMM(InputData=list(),
OmicsData=OmicsData,  OmicsDataMap=OmicsDataMap, trainID=trainID ,OmicsKernelMatrix=list(),
annotation=annotation, Y=Y, X=Expr, kernelsOmics =NA, AllRegions=0,
stdData = TRUE,
predict = predict, weight.fixed = weight.fixed, weight.random = weight.random,
maxiter = maxiter, outputall = outputall)

## Checking Pearson Correlations ##
cor(ResultTest$Ypred,ResultTest$outcome[!names(ResultTest$outcome) \%in\% trainID])
}
